craft_application/util/yaml.py

```diff
@@ -0,0 +1,79 @@
+# This file is part of craft_application.
+#
+# Copyright 2023 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify it
+# under the terms of the GNU Lesser General Public License version 3, as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranties of MERCHANTABILITY,
+# SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License along
+# with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""YAML helpers for craft applications."""
+from collections.abc import Hashable
+from typing import Any, Dict, Set, TextIO
+
+import yaml
+
+
+def _check_duplicate_keys(node: yaml.Node) -> None:
+    """Ensure that the keys in a YAML node are not duplicates."""
+    mappings: Set[yaml.Node] = set()
+
+    for key_node, _ in node.value:
+        try:
+            if key_node.value in mappings:
+                raise yaml.constructor.ConstructorError(
+                    "while constructing a mapping",
+                    node.start_mark,
+                    f"found duplicate key {key_node.value!r}",
+                    node.start_mark,
+                )
+            mappings.add(key_node.value)
+        except TypeError:  # pragma: no cover
+            # Ignore errors for malformed inputs that will be caught later.
+            pass
+
+
+def _dict_constructor(
+    loader: yaml.Loader, node: yaml.MappingNode
+) -> Dict[Hashable, Any]:
+    _check_duplicate_keys(node)
+
+    # Necessary in order to make yaml merge tags work
+    loader.flatten_mapping(node)
+    value = loader.construct_mapping(node)
+
+    try:
+        return dict(value)
+    except TypeError as type_error:
+        raise yaml.constructor.ConstructorError(
```

Comment:
Behold! A wild missingno has appeared!

https://github.com/canonical/craft-application/blob/24dc16736aead0ba53d0040dcd22272c81613501/tests/unit/util/invalid_yaml/unhashable.yaml-invalid

(Good catch and thank you for the good direction too!)

Reply:
Fun fact: that `except` clause appears to be unnecessary (codecov caught that), since the loader is raising that exception before the `_dict_constructor` gets to it. 

I've made #24 to kick this can down the road since I'm unsure.